scripts/source-maintenance.sh

```diff
@@ -232,10 +232,8 @@ for FILENAME in `git ls-files`; do
 	chmod 755 ${FILENAME}
 	;;
 
-    Makefile.am)
-
-    	perl -p -e 's/@([A-Z0-9_]+)@/\$($1)/g' <${FILENAME} >${FILENAME}.styled
-	mv ${FILENAME}.styled ${FILENAME}
+    *.am)
+		run_ applyPlugin scripts/sort-makefile-am-sources.pl ${FILENAME}
```

Comment:
Interesting... This is not what I meant when I [asked](https://github.com/squid-cache/squid/pull/643#discussion_r430077118) to use the new plugin interface. I meant converting your script into a plugin, to be automatically picked up by the existing applyPlugin code (instead of calling it explicitly). I _am_ glad that our applyPlugin function is so flexible that it can accommodate this new script like that, even though it was not written with this use case in mind.

I understand that the existing applyPlugin-calling code is only applied to C++ files, not Automake files. Can we extend that existing code to apply to Automake files as well somehow? Perhaps by adding a category prefix to plugin filenames? Or recording that category inside the plugin script itself using some simple markup?? If nothing like that is possible or desirable, then I am OK with this "hard-coded plugin" approach, but it would be nice to avoid this hard-coding for non-C++ files!

Reply:
TBH I don't think we want to. The plugins interface is meant with a specific use case in mind; if we adapt it to more than one use case we need to have a "negotiation" feature to say that some plugins apply to c/c++, some others to *.am, and from there on it's all downhill. How do we handle the case of a plugin applying to more than one category? The approaches you suggest all work, but I'd keep it simple until there is a clear benefit from having a class of plugins instead of this simple single one.

Reply:
> The plugins interface is meant with a specific use case in mind

That interface was built using a specific use case, but, in my mind, this use case (and others) are fundamentally the same. If we can use a unified approach for formatting C++, Automake, and shell files (to name a few categories), we should consider it.

> we need to have a "negotiation" feature to say that some plugins apply to c/c++, some others to *.am, and from there on it's all downhill.

... and the existing switch statement already does that, essentially. The requirement to negotiate, on its own, cannot be used as a proof that the plugin interface should be limited to C++ sources.

> How do we handle the case of a plugin applying to more than one category?

The existence of this question, on its own, cannot be used as a proof that the plugin interface should be limited to C++ sources. One possible answer would be that an am/plugin and cpp/plugin can both call the same general script.

There are many ways to add support for different source categories. Some good, some bad. Some may require complex negotiations. Some may require none. It is up to us to solve this puzzle. It is possible that solving it is not worth it, but it would be nice to understand why that is. After all, we did decide that adding a plugin interface was the right way forward in a C++ context. How is Automake context different?

> I'd keep it simple until there is a clear benefit from having a class of plugins instead of this simple single one.

Well, there is already a benefit, but, more importantly, I am worried that the author of the future PR adding the second Automake script will reject requests to "pluginize" it because the "existing code" does not do it. We can try to mitigate that risk by adding a comment:

```shell
run_ applyPlugin scripts/sort-makefile-am-sources.pl ${FILENAME}
# if there is a second script, "pluginize" both instead of hard-coding each
```

Reply:
added comment

Reply:
While sleeping on it, I realized that there is a simple solution to this problem that does not increase complexity while fully addressing my primary concern. I would like to commit that change (it cannot be expressed as a single suggestion due to GitHub limitations) for your review. Please let me know when it is a good time to do so.

FWIW, I recalled that the recently added plugin API is not so much about the automated selection of plugins -- that is actually a minor or even questionable benefit. Its primary accomplishment is a unified application of plugins. This PR has already demonstrated how easy it is to lose the hard-won advantages of the original feature: We nearly lost the fail-on-error behavior (restored in https://github.com/squid-cache/squid/commit/11711cb16609ba2d48294d2364f7855ba7bfeecf#diff-3c40c1f2e37a3a58de4a36e12819117e). I suspect there is at least one more subtle difference between the current PR code and the official plugin invocation code. It too will be gone with the upcoming change I want to share with you.

Reply:
I think this is the last old unresolved change request. @kinkie, please let me know when it is a good time to apply my changes for your review.

Reply:
> I think this is my last old unresolved change request. @kinkie, please let me know when it is a good time to apply my changes for your review.

Can it be done as a followup PR? That would be the easiest path forward.

Reply:
The best path forward for me would be to do it in this PR. You can easily revert the commit if you do not like it -- I can even do that reversal for you if needed. Needless to say, I will check that my commit does not change the script result/output.

However, if you insist, I will try to remember to create a new PR after this is merged. Your call.

Reply:
I committed my suggestion after checking that it does not change script's outcome (in my environment). Please review 55bbc14.